Servers/UI/OJS.Servers.Ui/ClientApp/src/components/guidelines/trees/Tree.tsx

```diff
@@ -1,97 +1,125 @@
+    onSelect: (node: ITreeItemType) => void;
     defaultSelected?: string;
     defaultExpanded?: string[];
+    itemFunc?: (item: ITreeItemType) => React.ReactElement;
+    treeItemHasTooltip?: boolean;
 }
 
 const Tree = ({
     items,
-    onTreeLabelClick,
-    defaultSelected,
+    onSelect,
+    defaultSelected = '',
     defaultExpanded = [],
-} : ITreeProps) => {
-    const [ expanded, setExpanded ] = useState([] as string[]);
-    const [ selected, setSelected ] = useState('');
+    itemFunc,
+    treeItemHasTooltip = false,
+}: ITreeProps) => {
+    const [ expandedIds, setExpandedIds ] = useState([] as string[]);
+    const [ selectedId, setSelectedId ] = useState('');
 
     const handleTreeItemClick = useCallback(
         (node: ITreeItemType) => {
             const id = node.id.toString();
-            let newExpanded = Array.from(expanded);
-
-            if (expanded.includes(id)) {
-                newExpanded = newExpanded.filter((e) => e !== id);
-            } else {
-                newExpanded.push(id);
-            }
+            const newExpanded = expandedIds.includes(id)
+                ? without(expandedIds,id)
+                : [ ...expandedIds,id ];
 
-            setExpanded(newExpanded);
+            setExpandedIds(newExpanded);
         },
-        [ expanded, setExpanded ],
+        [ expandedIds, setExpandedIds ],
     );
 
     const handleLabelClick = useCallback(
         (node: ITreeItemType) => {
-            setSelected(node.id.toString());
+            setSelectedId(node.id.toString());
 
-            onTreeLabelClick(node);
+            onSelect(node);
         },
-        [ onTreeLabelClick ],
+        [ onSelect ],
+    );
+
+    const renderTreeItem = useCallback((node: ITreeItemType) => (
+        <TreeItem
+            className={styles.treeElement}
+            key={node.id}
+            nodeId={node.id.toString()}
+            label={node.name}
+            onClick={() => handleTreeItemClick(node)}
+            onLabelClick={() => handleLabelClick(node)}
+        >
+            {isArray(node.children)
+                ? node.children.map((child) =>
+                    treeItemHasTooltip
+                        ? (<div className={styles.categoriesTree} key={child.id}>
+                            <div className={styles.tooltip}>
+                                {child.name}
+                            </div>
+                            {renderTreeItem(child)}
+                        </div>)
+                        : renderTreeItem(child))
+                : null}
+        </TreeItem>
+    ),[ handleLabelClick, handleTreeItemClick, treeItemHasTooltip ]);
+
+
+
+    const defaultItemFunc = useCallback(
+        (node: ITreeItemType) => 
+            treeItemHasTooltip
+                ? (<div className={styles.categoriesTree} key={node.id}>
+                    <div className={styles.tooltip}>
+                        {node.name}
+                    </div>
```

Comment:
Can be extracted as a render function in order to re-use it above(line 63). 
Is this style styles.categoriesTree correctly named?